### DISCUSSION TITLE
Web REPL for pico w

### DISCUSSION THREAD

**Discussion by smatterchoo** [micropython/micropython#10238] [RP2040 / Pico]
I'm eagerly awaiting release of this feature for the pico w.
Is there a beta channel I could try and maybe help with testing?
(To be clear, I'm referring to the updated 1.20 firmware for micropython which I have read will support the web repl server on the pico, not the desktop web repl client .js software)

**Comment by robert-hh** [comment]
Try the actual daily build, which is very close to v1.20. Webrepl shoul be enabled there.

**Reply by smatterchoo** [reply]
Thanks but that didn't work.
I grabbed the latest build and was unable to import 

`MicroPython v1.19.1-776-gbb77c1d5a on 2022-12-16; Raspberry Pi Pico W with RP2040`
`>>> import webrepl_setup`
`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'webrepl_setup'`
`>>> import webrepl`
`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'webrepl'`

**Comment by robert-hh** [comment]
You can get these to scripts from the MP web site: https://github.com/micropython/micropython-lib/tree/master/micropython/net/webrepl
Copy these two scripts to your board, e.g. with mpremote.py, and then you can use them.

**Reply by smatterchoo** [reply]
Thanks!  I'm having an issue with the terminal widget.  
I'm able to type my password and connect, and it gives me a `>>>` prompt and a cursor. But after that, all keystrokes seem to be ignored.  
I tried clicking on another element (which makes the terminal cursor disappear) and then clicking back in the terminal (making the cursor appear again), but still nothing happens when I type. 
FWIW I am able to move files between the pico w and my desktop via the webrepl page. 
To be clear I'm loading the page that is served directly by the pico, not the one hosted at micropython.org.
Also note the chrome javascript console for this local webrepl page flags that many scripts are being loaded from micropython.org and may be blocked in the future.

**Reply by petermoscickidev** [reply]
I tried to replicate your process, but I only was able to make the setup work. Did pico create a wifi network right after setup? Did you use webrepl.start()? Can you provide more info so I can replicate your process, step by step?

**Reply by smatterchoo** [reply]
thanks! yes, I did webrepl.start() after downloading and importing the files robert-hh mentioned.

webrepl.start() returns:
WebREPL server started on http://192.168.0.173:8266/
Started webrepl in normal mode

visiting that URL in chrome gives me a webpage with the expected elements. 
I can click the "connect" button next to the websocket URL ws://192.168.0.173:8266

the terminal widget then allows me to type the webrepl password, and then shows:
`Welcome to MicroPython!`
`Password:                   `                                                                                                                          
`WebREPL connected`                                                                                                                         
`>>> `   

After the `>>>` I was expecting to be able to type, but all keystrokes are ignored.

but the file upload/download widgets work.  

OK interesting.. I tried Control-D in the terminal, and when I did that, the characters that I had previously typed were then printed in the terminal.  So I can type `print('a')` in the term window, and I can't see any of those characters, but if I then hit Control-D, the character 'a' will be printed

It feels like maybe the "hide keystrokes" mode during typing the password isn't getting reset to "show keystrokes" mode for regular REPL interaction?

This is in chrome Version 108.0.5359.125 (Official Build) (64-bit) on windows 

The chrome javascript console has several warnings:
webrepl_content.js:2 A parser-blocking, cross site (i.e. different eTLD+1) script, https://micropython.org/webrepl/term.js, is invoked via document.write. The network request for this script MAY be blocked by the browser in this or a future page load due to poor network connectivity. If blocked in this page load, it will be confirmed in a subsequent console message. See https://www.chromestatus.com/feature/5718547946799104 for more details.

**Reply by shamlian** [reply]
I am having this same issue being unable to see what I've typed, but having the program run when I hit ^D, in Micropython 1.22.0 on a Raspberry Pi Pico W and using Chrome 120.0.6099.130, when connected over a webrepl.

**Comment by jimmo** [comment]
@smatterchoo It's on the TODO list for the v1.20 release to ensure that all network-capable boards get webrepl by default. But this hasn't been done yet, so you either need to install teh files manually (or easier, by using `mip`).

I will try and replicate the issue with the keystrokes.

**Reply by petermoscickidev** [reply]
>I will try and replicate the issue with the keystrokes.

That would be great, I'm experiencing exactly the same issue (with ctrl-d, enter not working)
